lib/moab/storage_object_validator.rb

```diff
@@ -0,0 +1,198 @@
+require 'moab'
+
+# Shameless gree: repetitious code included.
+module Moab
+  # Given a druid path, are the contents actually a well-formed Moab?
+  class StorageObjectValidator
+
+    EXPECTED_VERSION_SUB_DIRS = ["data", "manifests"].freeze
+    EXPECTED_DATA_SUB_DIRS = ["content", "metadata"].freeze
+    IMPLICIT_DIRS = ['.', '..'].freeze # unlike Find.find, Dir.entries returns these
+
+    # error codes
+    INCORRECT_DIR = 0
+    MISSING_DIR = 1
+    EXTRA_CHILD_DETECTED = 2
+    EMPTY = 3
+    NO_SIGNATURE_CATALOG = 4
+    NO_MANIFEST_INVENTORY = 5
+    NO_XML_FILES = 6
+    VERSIONS_NOT_IN_ORDER = 7
+    FILES_IN_VERSION_DIR = 8
+    VERSION_DIR_BAD_FORMAT = 9
+
+    ERROR_CODE_TO_MESSAGES = {
+      INCORRECT_DIR=> "Incorrect items in path",
+      MISSING_DIR => "Missing directory: %{addl}",
+      EXTRA_CHILD_DETECTED => "Unexpected item in path: %{addl}",
+      EMPTY => "No items in path",
+      FILES_IN_VERSION_DIR => "Top level should contain only sequential version directories. Also contains files: %{addl}",
+      NO_SIGNATURE_CATALOG => "Version: %{addl} Missing signatureCatalog.xml",
+      NO_MANIFEST_INVENTORY => "Version: %{addl} Missing manifestInventory.xml",
+      NO_XML_FILES => "Version: %{addl} Missing all required metadata files",
+      VERSIONS_NOT_IN_ORDER => "Should contain only sequential version directories. Current directories: %{addl}",
+      VERSION_DIR_BAD_FORMAT => "Version directory name not in 'v00xx' format"
+    }.freeze
+
+    attr_reader :storage_obj_path
+
+    def initialize(storage_object)
+      @storage_obj_path = storage_object.object_pathname
+      @directory_entries_hash = {}
+    end
+
+    def validation_errors
+      errors = []
+      errors.concat check_correctly_named_version_dir
+      errors.concat check_sequential_version_dirs if errors.empty?
+      errors.concat check_correctly_formed_moab if errors.empty?
+      errors
+    end
+
+    # TODO: Figure out which methods should be public
+
+    private
+
+    def check_correctly_named_version_dir
+      errors = []
+      sub_dirs(storage_obj_path).each do |version_dir|
+        if version_dir =~ /^[v]\d{4}/
+          nil
+        else
+          errors << result_hash(VERSION_DIR_BAD_FORMAT)
+        end
+      end
+      errors
+    end
+
+    def version_directories
+      @vdirs ||= sub_dirs(storage_obj_path)
+    end
+
+    def check_sequential_version_dirs
+      errors = []
+
+      version_directories.each_with_index do |dir_name, index|
+        expected_vers_num = index + 1 # version numbering starts at 1, array indexing starts at 0
+        begin
+          if dir_name[1..-1].to_i != expected_vers_num
+            errors << result_hash(VERSIONS_NOT_IN_ORDER, version_directories)
+            break
+          end
+        rescue ArgumentError
```

Comment:
I'm still not clear on how you would get an argument error, nor do you seem to be, as you don't have a test for this rescue clause, per coveralls.

Reply:
Ah yeah tomorrow you'll have to show me how to use coveralls to see what parts of my code is not tested and is tested. You get an ArgumentError when something cannot be converted to an integer from ``` if dir_name[1..-1].to_i != expected_vers_num ```

Reply:
So i'll change the error code to `VERSION_DIR_BAD_FORMAT` (used to make sure the version directories are correctly named)

Reply:
Switched it from  ``` to_i ``` to ``` Integer() ```